packages/nouns-contracts/test/foundry/DAOUpgrade/DAOUpgradeMainnetFork.t.sol

```diff
@@ -0,0 +1,185 @@
+    }
+
+    function propose(
+        address target,
+        uint256 value,
+        string memory signature,
+        bytes memory data,
+        uint32 clientId
+    ) internal returns (uint256 proposalId) {
+        vm.prank(proposerAddr);
+        address[] memory targets = new address[](1);
+        targets[0] = target;
+        uint256[] memory values = new uint256[](1);
+        values[0] = value;
+        string[] memory signatures = new string[](1);
+        signatures[0] = signature;
+        bytes[] memory calldatas = new bytes[](1);
+        calldatas[0] = data;
+        proposalId = NOUNS_DAO_PROXY_MAINNET.propose(targets, values, signatures, calldatas, 'my proposal', clientId);
+    }
+
+    function voteAndExecuteProposal(uint256 proposalId) internal {
+        NounsDAOTypes.ProposalCondensedV2 memory propInfo = NOUNS_DAO_PROXY_MAINNET.proposals(proposalId);
+
+        vm.roll(propInfo.startBlock + 1);
+        vm.prank(proposerAddr, origin);
+        NOUNS_DAO_PROXY_MAINNET.castRefundableVote(proposalId, 1);
+        vm.prank(WHALE, origin);
+        NOUNS_DAO_PROXY_MAINNET.castRefundableVote(proposalId, 1);
+
+        vm.roll(propInfo.endBlock + 1);
+        NOUNS_DAO_PROXY_MAINNET.queue(proposalId);
+
+        propInfo = NOUNS_DAO_PROXY_MAINNET.proposals(proposalId);
+        vm.warp(propInfo.eta + 1);
+        NOUNS_DAO_PROXY_MAINNET.execute(proposalId);
+    }
+}
+
+contract DAOUpgradeMainnetForkTest is DAOUpgradeMainnetForkBaseTest {
+    function setUp() public override {
+        super.setUp();
+
+        // Deploy the latest DAO logic
+        vm.setEnv('DEPLOYER_PRIVATE_KEY', '0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa');
+        newLogic = address(new NounsDAOLogicV4());
+
+        // Propose the upgrade
+        vm.setEnv('PROPOSER_KEY', '0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb');
+        vm.setEnv('DAO_V3_IMPL', Strings.toHexString(uint160(newLogic), 20));
+        vm.setEnv('PROPOSAL_DESCRIPTION_FILE', 'test/foundry/DAOUpgrade/proposal-description.txt');
+        uint256 proposalId = new ProposeDAOUpgradeMainnet().run();
+
+        // Execute the upgrade
+        voteAndExecuteProposal(proposalId);
+    }
+
+    function test_daoUpgradeWorked() public {
+        assertTrue(CURRENT_DAO_IMPL != NOUNS_DAO_PROXY_MAINNET.implementation());
+        assertEq(newLogic, NOUNS_DAO_PROXY_MAINNET.implementation());
+    }
+
+    function test_givenRecentBitPacking_creationBlockAndProposalIdValuesAreLegit() public {
+        NounsDAOTypes.ProposalCondensed memory prop = NOUNS_DAO_PROXY_MAINNET.proposalsV3(493);
+
+        assertEq(prop.id, 493);
+        assertEq(prop.creationBlock, 19093670);
+        assertEq(prop.creationTimestamp, 0);
+
+        prop = NOUNS_DAO_PROXY_MAINNET.proposalsV3(474);
+
+        assertEq(prop.id, 474);
+        assertEq(prop.creationBlock, 18836862);
+        assertEq(prop.creationTimestamp, 0);
```

Comment:
@eladmallel would be nice to also add a test that creationTimestamp is being written in new proposals